### ISSUE TITLE
Test documentation for broken links

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#4058]
It would be useful to have a test that walks through our documentation and fails if any links are broken.  There are probably standard ways to do this (some web searching may be the right place to start) but I suspect that a regex through all rst files in the docs/source directory followed by trying them all and looking at status messages would be an ok start.

**Comment by martindurant** [comment]
A regex that seems to do the job:
https://github.com/martindurant/filesystem_spec/blob/master/fsspec/implementations/http.py#L10

**Comment by theianrobertson** [comment]
Hey folks, I may have a bit of time to poke around with this - what would be your preference for how to do it?

1. Loop over some kind of regex on all RST files and try to hit all URLs in the test suite (would be faster, but wouldn't catch docstrings/dynamic links in the docs)
2. Run the sphinx linkcheck as above as part of the test suite, or even as a manual step in the docs build (may make things slow, but would catch links in the docstrings, and sphinx-generated links)
    - If this version is prefered, are docs auto-built?  I can't see anything but maybe I'm not looking hard enough.  If docs are auto-built maybe that's the right place to run this check, instead of in the unit tests?

**Comment by mrocklin** [comment]
I don't personally have enough knowledge to state a preference here.
Ideally whoever would do this would investigate both options and then
perhaps share some information, along with a proposal.

**Comment by JimCircadian** [comment]
Interestingly, this might factor into how we parse the links in the documentation: [Add support for rate limit detection / retries @ linkcheck
](https://github.com/sphinx-doc/sphinx/issues/7388)

Just had to kill the documentation linkchecking as there are too many links in the changelog. Going to look at excluding that and also evaluate the options discussed to look at testing the docs.

**Comment by JimCircadian** [comment]
Thankfully there's a useful `linkcheck_ignore = []` sphinx configuration option, so can utilise that for the moment.

**Comment by JimCircadian** [comment]
So this option works quite well. It ran in less than two minutes on my box so isn't onerous... Looking at the two options above:

1. Yes, more malleable to our needs but in danger of reinventing the wheel given sphinx provides a native approach. 
2. Ease of use, obviously requires a little investigation to get full coverage which I may raise as a separate issue.

The rate-limitations against github are an issue in both cases which is unfortunate, perhaps that is a separate issue to overcome via a set of cached results/leveraging the API to assess whether the underlying PRs/issues etc exist instead of hitting the https frontend. 

In the meantime I'm going to fix the following links via ignores (localhost) or via an appropriate method: 

```
(line   96) broken    https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html - 404 Client Error: Not Found for url: https://numpy.org/doc/stable/reference/c-api.generalized-ufuncs.html
(line   73) broken    https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html - 404 Client Error: Not Found for url: https://numpy.org/doc/stable/reference/c-api.generalized-ufuncs.html
(line   82) broken    https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html - 404 Client Error: Not Found for url: https://numpy.org/doc/stable/reference/c-api.generalized-ufuncs.html
(line    4) broken    https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html - 404 Client Error: Not Found for url: https://numpy.org/doc/stable/reference/c-api.generalized-ufuncs.html
(line    4) broken    https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html#details-of-signature - Anchor 'details-of-signature' not found
(line  107) broken    https://distributed.dask.org/en/latest/api.html#distributed.client.Client.persist - Anchor 'distributed.client.Client.persist' not found
(line   95) broken    https://pandas.pydata.org/pandas-docs/stable/timeseries.html#partial-string-indexing - Anchor 'partial-string-indexing' not found
(line  158) broken    https://distributed.dask.org/en/latest/web.html - 404 Client Error: Not Found for url: https://distributed.dask.org/en/latest/web.html
(line  158) broken    https://distributed.dask.org/en/latest/web.html - 404 Client Error: Not Found for url: https://distributed.dask.org/en/latest/web.html
(line  233) broken    https://distributed.dask.org/en/latest/local-cluster.html - 404 Client Error: Not Found for url: https://distributed.dask.org/en/latest/local-cluster.html
(line   33) broken    http://localhost:8787/status - HTTPConnectionPool(host='localhost', port=8787): Max retries exceeded with url: /status (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f6a02170978>: Failed to establish a new connection: [Errno 111] Connection refused',))
(line   47) broken    http://localhost:8787 - HTTPConnectionPool(host='localhost', port=8787): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f6a24c48cf8>: Failed to establish a new connection: [Errno 111] Connection refused',))
(line  131) broken    https://arrow.apache.org/docs/python/filesystems.html#hadoop-file-system-hdfs - Anchor 'hadoop-file-system-hdfs' not found
(line  183) broken    https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#shared-credentials-file - Anchor 'shared-credentials-file' not found
(line   23) broken    http://localhost:8787/status - HTTPConnectionPool(host='localhost', port=8787): Max retries exceeded with url: /status (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f6a02170978>: Failed to establish a new connection: [Errno 111] Connection refused',))
(line  188) broken    https://jcrist.github.io/jupyterhub-on-hadoop/ - 404 Client Error: Not Found for url: https://jcristharif.com/jupyterhub-on-hadoop/
(line   78) broken    https://www.anaconda.com/support - 404 Client Error: Not Found for url: https://www.anaconda.com/support
```

@mrocklin do you have any feedback on this - is it worth opening a investigating/opening a separate issue for the github checks?

**Comment by gforsyth** [comment]
Looks like https://github.com/jupyterlab/pytest-check-links has exponential backoff, might be worth investigating as an option

**Comment by JimCircadian** [comment]
Commented on the PR to see whether it's worth merging the link fixes first, but I should be able to pick that up and explore options later this week.

@gforsyth I couldn't see any mention of rate limiting / backoff in that pytest library or the requests_cache layer underneath. Do you have a link to where you saw that? Would be good to try it out if it's there. 

It does have a caching layer which is kind of interesting, but I don't think with the amount of github links in the changelog that would help (github throttles within one check run for me). However, I'm happy to have a bit more of a look around if people think it's worth overcoming this problem!!! :-)

**Comment by gforsyth** [comment]
Hey @JimCircadian -- I misread it the first time, but it does have logic specific to handling GitHub `Retry-After` timings: https://github.com/jupyterlab/pytest-check-links/pull/7

**Comment by JimCircadian** [comment]
Check that out! Thanks @gforsyth, totally missed it. I'll have a crack at leveraging that this coming weekend! :+1:

**Comment by JimCircadian** [comment]
So it ran well and responded to the github retry-wait header nicely, resulting in a run of ~50 minutes.

Ignoring the localhost errors, the failures picked up were probably worthwhile given there are some dead github links (I'll fix / remedy those where I can.) The question is, should this be an automated test? Happily have updated the makefile for docs and am playing with the link caching facility for py.test, I would wager the additional coverage of github is worth it from the results. 

Thoughts / comments welcome!

`
________________________ <redacted>/code/public/dask/docs/source/caching.rst: diagnostics.rst ________________________
diagnostics.rst: No such file: <redacted>/code/public/dask/docs/source/diagnostics.rst
_________________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/justinwp _________________
https://github.com/justinwp: 404: Not Found
______________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/daniel-severo _______________
https://github.com/daniel-severo: 404: Not Found
_________________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/justinwp _________________
https://github.com/justinwp: 404: Not Found
______________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/daniel-severo _______________
https://github.com/daniel-severo: 404: Not Found
_________________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/jmmease __________________
https://github.com/jmmease: 404: Not Found
_________________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/jmmease __________________
https://github.com/jmmease: 404: Not Found
_________________ <redacted>/code/public/dask/docs/source/changelog.rst: https://github.com/jmmease __________________
https://github.com/jmmease: 404: Not Found
_ <redacted>/code/public/dask/docs/source/configuration.rst: https://docs.python.org/3/library/ast.html#ast.literal_eval _
https://docs.python.org/3/library/ast.html#ast.literal_eval: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
`

**Comment by jsignell** [comment]
I wonder if it'd make sense to run this test only right before doc build.

**Comment by gforsyth** [comment]
> I wonder if it'd make sense to run this test only right before doc build.

Like just before a release?  That makes sense to me.

**Comment by jsignell** [comment]
> Like just before a release? 

Yeah, I guess that could be pretty annoying to get blocked on, but it would force us to make sure that the links are uptodate before publishing. The other options that I see are run this test on a cron job (nightly or something), but in my experience failures in tests that run on cron jobs are easy to ignore.

**Comment by gforsyth** [comment]
It might be a little annoying, but also the rate of link-rot probably isn't _that_ high.  And strong agree on nightly failures and the propensity for them to be ignored.

@jrbourbeau -- as someone who does very many dask releases, does this seem like an unreasonable ask?

**Comment by jrbourbeau** [comment]
It looks like we're ignoring GitHub / localhost links over in https://github.com/dask/dask/pull/6403. I ran `make linkcheck` with the changes in https://github.com/dask/dask/pull/6403 and it look ~2 minutes to run on my laptop. Given the time it takes our tests to run, I'd be inclined to add `make linkcheck` to our documentation build in CI. Though please correct me if there's some other reason to not include link checking in CI 

> I wonder if it'd make sense to run this test only right before doc build

Currently our docs are built and re-deployed with every merge to `master`

**Comment by jsignell** [comment]
> Currently our docs are built and re-deployed with every merge to master

Ah, I forgot about that. Maybe we could add a test that just runs when there are changes to the docs? But if that is too much fuss I'm really ok with whatever. My main worry is that if there is ever a failure in these tests it will almost certainly not be related to the work in a particular PR. So it'll only be a source of frustration for contributors.

**Comment by JimCircadian** [comment]
Hello all

So I've left the configuration changes in for using sphinx-build but updated the makefile and doc dependencies to use pytest as it provides the best coverage. 

However, perhaps for CI purposes we can use a separate sphinx based command which skips over the github links that cause the issues. 

I've also updated the docs to fix all the broken links as they currently exist. Just having a bit of trouble with the CI testing as that's failing (though I don't think that's related to my changes). 

I'll assume (unless there's more feedback) that it's worth concluding that PR, adding an additional makefile command for CI purposes is trivial @jrbourbeau though I'd think that automatically checking them is less valuable without good coverage than manually checking them on a release, with complete coverage.

Hope that helps!!!

**Comment by ncclementi** [comment]
Hi folks, checking in here, it seems this was addressed here https://github.com/dask/dask/pull/6403 and we could probably close this issue, is that accurate or am I missing something? 
cc: @jsignell

**Comment by jakirkham** [comment]
Agreed let's close. That PR both fixed broken links and added a CI check. If there's something else missing, please feel free to reopen.

**Comment by jsignell** [comment]
Thanks @JimCircadian for your work on this!
